### ISSUE TITLE
Slots 16,833,545 through 16,903,261 are missing from google bigtable

### ISSUE THREAD

**Issue by bji** [solana-labs/solana#32006]
#### Problem

These slots are not there at all.  ledger_storage.get_confirmed_blocks() for any slots in that range returns nothing.  Also, solana explorer behaves oddly at slot 16,833,544 -- it shows no child slot.  And at slot 16,903,262 it shows a parent slot of 16,903,261 but following the link shows a failure page.

#### Proposed Solution

If you have the means to upload these slots to google bigtable, please do.

**Comment by steviez** [comment]
I found and downloaded the ledger backup for this range. I initially opened with `solana-ledger-tool` from v1.1.18 (compiled with rust 1.44.0), but was also able to open it with `v1.16.0` after passing `--force-update-to-open` once. Running the below command, I see the following (noting that the command list roots in descending order):
```
$ solana-ledger-tool list-roots --ledger ~/gh32006/ --start-root 16833544 --num-roots 5
[2023-06-07T06:15:23.504275472Z INFO  solana_ledger_tool] solana-ledger-tool 1.16.0 (src:e0fcdbb0; feat:2891131721, client:SolanaLabs)
...
16903265: 25S5iAUH5gKsri7W15BHxCdytAXsr7WkTm5MSskYHpaR
16903264: 8pPnwpiggMDHqQYqhALxKt1RgEgWaUyjHsNnQAC49yYQ
16903263: FkHgEnRY9Fi6obJHSpaoiDaArWrMF1wPqyiGteQ1qx1z
16903262: FWYGZ1uQnqMwnbGLkWf4boZCLfTGTpMKWxGAy9YUan22
16833544: Fe9yXKpTrZxydZDCntYpvp8Povp7gucTT58HbJMz5tya
```
This command just reads the output of roots column, but it would suggest that the gap you mentioned was indeed a gap. However, looking at some individual blocks,
```
$ solana-ledger-tool slot 16833544 16903262 --ledger ~/gh32006/ -v
    SlotMeta { slot: 16833544, ..., parent_slot: Some(16833543), next_slots: [16833545], ... }

    SlotMeta { slot: 16903262, ..., parent_slot: Some(16903261), next_slots: [16903263], ... }
```
The output from the individual slots would indicate there are more slots in the chain. So as you can see, the output from these two commands are contradictory. The blocks actually have the votes so they are the source of truth. The node that generated the backup was likely the same one loading up to bigtable, so if it didn't identify any roots in that range, that would also likely explain why they aren't in bigtable.

In order to properly figure this out, I think we will probably need to replay those ~70k slots with the original version (v1.1.18). Doing so will also likely entail some manual checks to ensure that what we compute is good; should be able to do this by comparing the bank hashes to hashes embedded in votes in the blockstore.

v1.1.18 doesn't have the code to set roots or store tx status in ledger-tool codepath by default, so this is going to take some effort

**Comment by CriesofCarrots** [comment]
Isn't this the usual (legacy) issue with missing root entries in the root column? Do you have reason to suspect the parent-child chain back from the first root after the gap isn't correct?
At a glance, I don't think we need to actually replay anything; just backport the repair-roots code to v1.1.18. We've done this for other old versions. I can probably pick this up (cc @willhickey )

**Comment by steviez** [comment]
> Isn't this the usual (legacy) issue with missing root entries in the root column? 

Uhh, probably; I'm not familiar with this issue.

> Do you have reason to suspect the parent-child chain back from the first root after the gap isn't correct?

I do not

> At a glance, I don't think we need to actually replay anything; just backport the repair-roots code to v1.1.18. We've done this for other old versions.

Awesome, I'm not familiar with this.

> I can probably pick this up

Even better! However, that may not even be necessary. I downloaded the backup from one of the other warehouse nodes, and that version seems to already have the roots in proper order.
```
$ solana-ledger-tool list-roots --ledger ./ --start-root 16833544 --num-roots 10
...
16833555: 7ifAKbh5NT12D9K3zyub4EjaYineSoVS51CmMH6W66xb
16833554: UAek48GFPiiRVLb1PxLRauEFJhephwWUpAPZUph3XGB
16833553: 139pQ3RQq8n51WvvW3VKguUFMQmAphVqR3B5tdKBkJ3G
16833552: 8JKAeoYnuLscKzLMMQ6nRgFuz9e6TUk9SVF2GmMTMtWm
16833551: ByokGc5A7Dy34r7zUnsoyekbUDZSe3ZFty1Gc8cscFyi
16833550: ATpKFB9i8QBQXbqZVLeDtBLFVFsd2NdfexXrjT9txBt4
16833549: 8kzaFL7JAwzpvXkP7ks9XU3D1KSVVsEWAwwYDpmdLBJs
16833546: CzdDM5Jj7b1rHc1rmhWs9CiPQxzcscNUf2MQtJcRJSvN
16833545: DRC5kmBASqDAJ3nkdv3EzSwGCfUipXLtneirnJdi4dFE
16833544: Fe9yXKpTrZxydZDCntYpvp8Povp7gucTT58HbJMz5tya
```
and it also has `16903262` rooted with same hash as from the other node:
```
$ solana-ledger-tool list-roots --ledger ./ --start-root 16903262 --num-roots 1
...
16903262: FWYGZ1uQnqMwnbGLkWf4boZCLfTGTpMKWxGAy9YUan22
```

**Comment by CriesofCarrots** [comment]
> So, this second ledger does NOT appear to have the issue you mentioned, so believe we can just do a straight forward upload using it as the source?

Yep! Seems like it. I would expect us to need to use the v1.1.18 binary for the upload because of changes to transaction metadata. Are you keen to give it a go, @steviez ?

**Comment by steviez** [comment]
I ended up picking this up with guidance from CriesofCarrots. To leave a paper trail incase we have to do this again in the future:
- I started by creating a branch from the v1.1.18 tag and confirmed I could build as-is with `cargo +1.44.0`
- I backported two commits from v1.2 to the branch, and then made necessary changes to get it compiling. The last commit was mostly commenting out stuff that did not exist in v1.1 in order to avoid having to backport even more code.
    - [0001-Long-term-ledger-storage-with-BigTable-bp-11222.patch](https://github.com/solana-labs/solana/files/11726864/0001-Long-term-ledger-storage-with-BigTable-bp-11222.patch)
    - [0002-Cargo.lock.patch](https://github.com/solana-labs/solana/files/11726865/0002-Cargo.lock.patch)
    - [0003-Get-ledger-tool-building-with-bigtable-command-backp.patch](https://github.com/solana-labs/solana/files/11726866/0003-Get-ledger-tool-building-with-bigtable-command-backp.patch)
- I extracted the ledger from `mainnet-beta-ledger-europe-fr2`; this backup did not have the missing-roots-column issue that the backup on `mainnet-beta-ledger-us-ny5` did. Having two sources (and one that didn't exhibit the issue) saved us the `repair-roots` step that CriesofCarrots mentioned.
- At this point, a straight forward run of `solana-ledger-tool bigtable upload ...` for the missing slot range

I did some spot checks (especially on the boundary slots) and things look to be in order now on explorer so closing this one out. Thanks for finding / reporting @bji !
